Language/Haskell/GhcMod/ImportedFrom.hs

```diff
@@ -0,0 +1,785 @@
+
+    showDebugStage "stage0" stage0
+
+    let stage1 = refineAs mySymbol stage0
+    showDebugStage "stage1" stage1
+
+    let stage2 = refineRemoveHiding stage1
+    showDebugStage "stage2" stage2
+
+    let stage3 = refineExportsIt symbolToUse stage2
+    showDebugStage "stage3" stage3
+
+    let stage4 = refineLeadingDot mySymbol stage3
+    showDebugStage "stage4" stage4
+
+    stage5 <- refineVisibleExports (ghcPkgHaddockInterface ghcPkg readProc pkgDbStack) stage4
+    showDebugStage "stage5" stage5
+
+    let lastMatch  = Safe.headMay $ catMaybes [getLastMatch stage5, getLastMatch stage4]
+
+    gmLog GmDebug "" $ strDoc $ show $ "guessHaddockUrl: lastMatch: " ++ show lastMatch
+
+    let lastMatchModule :: String
+        lastMatchModule = case mName <$> lastMatch of
+                            Just modn   -> modn
+                            _           -> error $ "No nice match in lastMatch for module: " ++ show lastMatch
+
+        lastMatchPackageName :: String
+        lastMatchPackageName = case mPackageName <$> lastMatch of
+                                Just p -> p
+                                _      -> error $ "No nice match in lastMatch for package name: " ++ show lastMatch
+
+    let getHaddockUrl = ghcPkgHaddockUrl ghcPkg readProc pkgDbStack :: String -> m (Maybe String)
+
+    haddock <- (maybe (return Nothing) getHaddockUrl . Just) lastMatchPackageName
+
+    gmLog GmDebug "" $ strDoc $ "guessHaddockUrl: lastMatchModule:      " ++ lastMatchModule
+    gmLog GmDebug "" $ strDoc $ "guessHaddockUrl: lastMatchPackageName: " ++ lastMatchPackageName
+    gmLog GmDebug "" $ strDoc $ "guessHaddockUrl: haddock:              " ++ show haddock
+
+    haddock <- return $ fromMaybe (error "haddock is Nothing :(") haddock
+
+    let f = haddock </> (moduleNameToHtmlFile lastMatchModule)
+
+    e <- liftIO $ doesFileExist f
+
+    if e then return $ Right $ "file://" ++ f
+         else do gmErrStrLn "Please reinstall packages using the flag '--enable-documentation' for 'cabal install.\n"
+                 return $ Left $ "Could not find " ++ f
+
+-- | Look up Haddock docs for a symbol.
+importedFrom
+    :: forall m. IOish m
+    => FilePath     -- ^ A target file.
+    -> Int          -- ^ Line number.
+    -> Int          -- ^ Column number.
+    -> Expression   -- ^ Expression (symbol)
+    -> GhcModT m String
+importedFrom file lineNr colNr (Expression symbol) = do
+    ghcPkg     <- getGhcPkgProgram
+    readProc   <- gmReadProcess
+    pkgDbStack <- getPackageDbStack
+
+    ghandle handler $
+      runGmlT' [Left file] deferErrors $
+        withInteractiveContext $ do
+          crdl   <- cradle
+          modSum <- fileModSummaryWithMapping (cradleCurrentDir crdl </> file)
+          let modstr = moduleNameString $ ms_mod_name modSum :: String
+
+          res <- guessHaddockUrl modSum file modstr symbol lineNr colNr ghcPkg readProc pkgDbStack
+
+          case res of Right x  -> return $ "SUCCESS: " ++ x ++ "\n"
```

Comment:
```
commit ae4badbb3984b5f718ff572f46d6b59c18d3a339
Author: Carlo Hamalainen <carlo@carlo-hamalainen.net>
Date:   Tue Jul 19 20:06:19 2016 +0800

    Use exit code instead of printing SUCCESS / FAIL.
```